sw/host/opentitanlib/src/test_utils/epmp.rs

```diff
@@ -0,0 +1,230 @@
+    /// Rule Locking Bypass: allow modification of locked ePMP entries.
+    pub const EPMP_MSECCFG_RLB: u32 = 4;
+}
+
+#[derive(Debug, Error)]
+pub enum EpmpError {
+    #[error("Invalid raw EPMP length (cfg={0}, addr={1})")]
+    InvalidLength(usize, usize),
+}
+
+/// Represents the different EPMP region kinds.
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum EpmpRegionKind {
+    Off = 0,
+    Tor = 1,
+    Na4 = 2,
+    Napot = 3,
+}
+
+impl Default for EpmpRegionKind {
+    fn default() -> Self {
+        EpmpRegionKind::Off
+    }
+}
+
+const EPMP_ADDR_SHIFT: u8 = 3;
+const EPMP_ADDR_MASK: u8 = 3;
+
+/// Represents the address range associated with the PMP entry.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct EpmpAddressRange(pub u64, pub u64);
+
+impl EpmpAddressRange {
+    pub fn start(&self) -> u64 {
+        self.0
+    }
+    pub fn end(&self) -> u64 {
+        self.1
+    }
+}
+
+/// A decoded ePMP entry.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct EpmpEntry {
+    /// The bit pattern present in `pmpNcfg`.
+    pub cfg: u8,
+    /// The region kind encoded ni the `cfg` field.
+    pub kind: EpmpRegionKind,
+    /// The address range for this entry.
+    pub range: EpmpAddressRange,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct Epmp {
+    pub entry: Vec<EpmpEntry>,
+}
+
+impl Epmp {
+    /// Creates an `Epmp` from the raw RISCV32 form of the ePMP.
+    pub fn from_raw_rv32(pmpcfg: &[u32], pmpaddr: &[u32]) -> Result<Self> {
+        ensure!(
+            pmpcfg.len() * 4 == pmpaddr.len(),
+            EpmpError::InvalidLength(pmpcfg.len(), pmpaddr.len())
+        );
+        let pmpcfg = pmpcfg
+            .iter()
+            .flat_map(|word| word.to_le_bytes())
+            .collect::<Vec<u8>>();
+        let mut entry = Vec::with_capacity(pmpaddr.len());
+        for i in 0..pmpaddr.len() {
+            let region_kind = match (pmpcfg[i] >> EPMP_ADDR_SHIFT) & EPMP_ADDR_MASK {
+                0 => EpmpRegionKind::Off,
+                1 => EpmpRegionKind::Tor,
+                2 => EpmpRegionKind::Na4,
+                3 => EpmpRegionKind::Napot,
+                _ => unreachable!(),
+            };
+            let addr = match region_kind {
+                EpmpRegionKind::Off => EpmpAddressRange(0, 0),
+                EpmpRegionKind::Tor => {
+                    if i == 0 {
+                        EpmpAddressRange(0, (pmpaddr[i] as u64) << 2)
+                    } else {
+                        EpmpAddressRange((pmpaddr[i - 1] as u64) << 2, (pmpaddr[i] as u64) << 2)
+                    }
+                }
+                EpmpRegionKind::Na4 => {
+                    EpmpAddressRange((pmpaddr[i] as u64) << 2, ((pmpaddr[i] + 4) as u64) << 2)
+                }
+                EpmpRegionKind::Napot => {
+                    let size = pmpaddr[i].trailing_ones();
+                    let addr = ((pmpaddr[i] & !((1 << size) - 1)) as u64) << 2;
```

Comment:
nit: this could be slightly simplified:
```suggestion
                    let addr = ((pmpaddr[i] & (pmpaddr[i] + 1)) << 2;
```

Reply:
Done.